src/protocols/rdp/rdp.c

```diff
@@ -227,10 +227,53 @@ static BOOL rdp_freerdp_authenticate(freerdp* instance, char** username,
 
     rdpContext* context = instance->context;
     guac_client* client = ((rdp_freerdp_context*) context)->client;
-
-    /* Warn if connection is likely to fail due to lack of credentials */
-    guac_client_log(client, GUAC_LOG_INFO,
-            "Authentication requested but username or password not given");
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    char* params[4] = {};
+    int i = 0;
+    
+    if (settings->username == NULL || strcmp(settings->username, "") == 0) {
+        params[i] = "username";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_USERNAME;
+        i++;
+    }
+    
+    if (settings->password == NULL || strcmp(settings->password, "") == 0) {
+        params[i] = "password";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_PASSWORD;
+        i++;
+    }
+    
+    if (settings->domain == NULL || strcmp(settings->domain, "") == 0) {
+        params[i] = "domain";
+        rdp_client->rdp_credential_flags |= GUAC_RDP_CRED_FLAG_DOMAIN;
+        i++;
+    }
```

Comment:
Can these ever be the empty string? It looks to me that argument parsing will ensure that an omitted parameter will be represented by `NULL`:

https://github.com/apache/guacamole-server/blob/c10ceab7e80f9d6eae9ad98254bf97f41a0de112/src/protocols/rdp/settings.c#L760-L773

If the concern here is an empty `argv` stream, perhaps the handling of the stream should ensure that it consistently assigns `NULL` for the relevant setting in those cases, such that code further downstream need not be aware of those concerns?

Reply:
Good point - and, there could be legitimate cases where users want an empty string versus a `NULL` value.  However, I'm not sure that there's currently a way to pass an empty string through??  But I think we can at least just simplify it to `NULL` checks and skip the empty string checks.

Reply:
For `settings->foo` coming from `guac_user_parse_args_string(..., NULL)`, yes, this will always be `NULL`. It would only perhaps be `""` if the call were `guac_user_parse_args_string(..., "")`.

BUT, it is likely possible for things to be `""` if the value is set via an empty `argv` stream. If that's the case, then I would suggest ensuring that handling of the `argv` stream uses `NULL` rather than `""` if it receives zero bytes overall.

Reply:
Got it - I've adjusted everything to behave that way.